src/adapter/src/catalog/builtin.rs

```diff
@@ -2371,6 +2371,67 @@ JOIN mz_catalog.mz_roles role_owner ON role_owner.id = class_objects.owner_id
 WHERE mz_schemas.database_id IS NULL OR d.name = pg_catalog.current_database()",
 };
 
+pub const PG_DEPEND: BuiltinView = BuiltinView {
+    name: "pg_depend",
+    schema: PG_CATALOG_SCHEMA,
+    sql: "CREATE VIEW pg_catalog.pg_depend AS WITH class_objects AS (
+    SELECT 
+        id,
+        CASE
+            WHEN type = 'table'  THEN 'pg_tables'::pg_catalog.regclass::pg_catalog.oid
+            WHEN type = 'source' THEN 'pg_tables'::pg_catalog.regclass::pg_catalog.oid
+            WHEN type = 'view'   THEN 'pg_views'::pg_catalog.regclass::pg_catalog.oid
+            WHEN type = 'materialized-view' THEN 'pg_matviews'::pg_catalog.regclass::pg_catalog.oid
+        END classid,
+        schema_id
+    FROM mz_catalog.mz_relations
+    UNION ALL
+    SELECT mz_indexes.id, mz_indexes.oid, schema_id
+    FROM mz_catalog.mz_indexes
+    JOIN mz_catalog.mz_relations ON mz_indexes.on_id = mz_relations.id
+),
+
+current_objects AS (
+    SELECT 
+        class_objects.id,
+        classid,
+        mz_objects.oid,
+        CASE
+            WHEN mz_catalog.mz_schemas.name = 'pg_catalog' THEN 'i' 
+            WHEN mz_catalog.mz_schemas.name = 'mz_catalog' THEN 'i' 
+            WHEN mz_catalog.mz_schemas.name = 'mz_internal' THEN 'i'
+            -- We mark sources as internal because they
+            -- are exposed via pg_catalog
+            -- as tables, but the system will error
+            -- if a user attempts DROP TABLE <source>.
+            WHEN (
+                SELECT count(*) FROM mz_catalog.mz_sources 
+                WHERE id = class_objects.id
+            ) > 0 THEN 'i'
+            ELSE 'n'
```

Comment:
As far as I can tell, this doesn't match PostgreSQL's documentation on what `i`/`DEPENDENCY_INTERNAL` means. The docs say:

> The dependent object was created as part of creation of the referenced object, and is really just a part of its internal implementation. A direct `DROP` of the dependent object will be disallowed outright (we'll tell the user to issue a `DROP` against the referenced object, instead). A `DROP` of the referenced object will result in automatically dropping the dependent object whether `CASCADE` is specified or not. If the dependent object has to be dropped due to a dependency on some other object being removed, its drop is converted to a drop of the referenced object, so that `NORMAL` and `AUTO` dependencies of the dependent object behave much like they were dependencies of the referenced object. Example: a view's `ON SELECT` rule is made internally dependent on the view, preventing it from being dropped while the view remains. Dependencies of the rule (such as tables it refers to) act as if they were dependencies of the view.

I think all dependencies in Materialize should be normal dependencies (we don't support `ON SELECT` rules on views), except for an index's dependency on its relation, which should be `a`/`DEPENDENCY_AUTO`.

Note, too, that the dependency type is a property of the dependency edge, not the object. So this logic would be better suited to the `SELECT` clause below, rather than being part of the `current_objects` CTE.

Reply:
@benesch so I’m happy to change the list. However, by marking sources as internal the external tools will not allow you to drop them. The reason I think that is important is we expose sources as tables, so those tools attempt drop table statements that will fail

Reply:
I’m speaking specifically of drop buttons in Postgres UIs that construct the DROP TABLE statement for the user

Reply:
> @benesch so I’m happy to change the list. However, by marking sources as internal then external tools will not allow you to drop them. The reason I think that is important is we expose sources as tables, so those tools attempt drop table statements that will fail

Hmm, what tool in what context? I'm confused about why a tool would care about what `pg_depend` says when going to drop a table. Wouldn't it just look at the entry for the table in `pg_class`, and not care about what the table depends on? I pointed DBeaver at your branch and it still lets me construct a `DROP TABLE` statement against the source. (Which then fails with a "cannot be run in transaction" error, ugh.)

Separately, the original sin here was pretending sources were tables, and I'm game for revisiting that decision. I'd rather not layer a hack on top of a hack, as some other tool will expect `pg_depend` to work as documented and will get real confused when it does not.

Say, it occurs to me: the dependency a progress source has on its parent source _is_ probably an internal dependency, as you can't drop the subsource independently of its parent source.

Reply:
I would be very game to make Kafka sources have a subsource/table, a la PG sources––and then make all subsources tables that can only be inserted into by the system.

Reply:
Same! Seems miserable to achieve without breaking backcompat, but I'd be very in favor of trying to figure out how to do it.